tensorrt_llm/llmapi/mm_encoder.py

```diff
@@ -0,0 +1,163 @@
+from pathlib import Path
+from typing import Any, Optional, Union
+
+from tensorrt_llm.inputs.data import PromptInputs
+from .llm import BaseLLM, _TorchLLM, RequestOutput
+from tensorrt_llm.sampling_params import SamplingParams
+from tensorrt_llm._utils import nvtx_range_debug
+from typing import List, Sequence
+from tensorrt_llm.inputs import prompt_inputs
+from tqdm import tqdm
+from tensorrt_llm.inputs import create_input_processor
+from .llm_args import PybindMirror
+from .mpi_session import external_mpi_comm_available
+from tensorrt_llm.bindings import executor as tllm
+
+class MultimodalEncoder(_TorchLLM):
+    """MultimodalEncoder class is the main class for running a multimodal encoder model using PyTorch backend.
+"""
+
+    def __init__(self,
+                 model: Union[str, Path],
+                 trust_remote_code: bool = False,
+                 tensor_parallel_size: int = 1,
+                 dtype: str = "auto",
+                 **kwargs: Any) -> None:
+
+        # Validate that users don't pass LLM-specific or TRT-specific arguments
+        self._validate_mm_args_for_torch_backend(kwargs)
+
+        super().__init__(model,
+                         trust_remote_code=trust_remote_code,
+                         tensor_parallel_size=tensor_parallel_size,
+                         dtype = dtype,
+                         **kwargs)
+
+    def _build_model(self):
+        BaseLLM._build_model(self)
+        assert self._engine_dir is None
+
+        # Tokenizer loading should be after calling model_loader(), since model_loader() may download the model from HF hub.
+        # It should also be before bindings ExecutorConfig, which may depend on tokenizer info.
+        self._tokenizer = self._try_load_tokenizer()
+
+        # Multimodal special handling:
+        # 1. Default load_tokenizer may fail because MM has different tokenizer configuration. Hence we initialize it inside input processor
+        # 2. May need to modify model weights for MM (e.g., resize vocab embedding). We must do such operation via input processor's __init__
+        self.input_processor = create_input_processor(self._hf_model_dir,
+                                                      self.tokenizer)
+        self.tokenizer = self.input_processor.tokenizer
+
+        max_batch_size = self.args.max_batch_size
+        max_num_tokens = self.args.max_num_tokens
+        max_seq_len = self.args.max_seq_len
+
+        kwargs = {}
+        if self._on_trt_backend:
```

Comment:
Is this encoder designed to work for both TRT and Pytorch? I am curious if both backend are need or we can only focus on the Pytorch path, the later one will simplify the code.